### PR TITLE
Add easyimage upload button in toolbar

### LINKED ISSUE #1730: Add a button for Easy Image plugin
## Are you reporting a feature request or a bug?

Feature request

## Provide detailed reproduction steps (if any)

In 4.9.0 we have added a Easy Image plugin.

While simplifying the feature we skipped the image button all along - the problem with this approach is that end user has no clear information whether dropping images (or EI plugin in general) is available at all.

The only way for end user is to drop the file and see if it's being uploaded. In the worst case scenario, page could be replaced with the dropped image.

The solution to that issue will be adding a button that will display a regular `input[file]` allowing for image upload.

## Other details

* CKEditor version: 4.9.0
* Installed CKEditor plugins: easyimage

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?
yes
### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?
- There is added new button to toolbar to group `insert`, as this group might contain similar types of buttons.
- button activate command which simulate clicking into `<input type="file">` element, and listen on change event. If user select some files, it simulate paste event.
- Input is not added to the DOM, but still might be clicked with a command.
- there is not too much unit test, as simulate pasting is already implemented in other tests. I haven't found a way to mock opening browsers' file managers, that's why I only detect existence of a command.

close #1730

### CODE REVIEW

**Comment by wwalc** [comment]
Just $0.02: `EasyimageUploadFile` button name looks a bit invalid. The feature supports uploading images, not files. How about `EasyImageUpload`? Or just `EasyImage`?

**Comment by wwalc** [comment]
Btw. note that there is inconsistency with the naming convention of other toolbar buttons (UpperCamelCase is used everywhere):

```
config.toolbar = [
		{ name: 'document', items: [ 'Source', '-', 'Save', 'NewPage', 'Preview', 'Print', '-', 'Templates' ] },
		{ name: 'clipboard', items: [ 'Cut', 'Copy', 'Paste', 'PasteText', 'PasteFromWord', '-', 'Undo', 'Redo' ] },
		{ name: 'editing', items: [ 'Find', 'Replace', '-', 'SelectAll', '-', 'Scayt' ] },
		{ name: 'forms', items: [ 'Form', 'Checkbox', 'Radio', 'TextField', 'Textarea', 'Select', 'Button', 'ImageButton', 'HiddenField' ] },
		'/',
		{ name: 'basicstyles', items: [ 'Bold', 'Italic', 'Underline', 'Strike', 'Subscript', 'Superscript', '-', 'CopyFormatting', 'RemoveFormat' ] },
		{ name: 'paragraph', items: [ 'NumberedList', 'BulletedList', '-', 'Outdent', 'Indent', '-', 'Blockquote', 'CreateDiv', '-', 'JustifyLeft', 'JustifyCenter', 'JustifyRight', 'JustifyBlock', '-', 'BidiLtr', 'BidiRtl', 'Language' ] },
		{ name: 'links', items: [ 'Link', 'Unlink', 'Anchor' ] },
		{ name: 'insert', items: [ 'Image', 'Flash', 'Table', 'HorizontalRule', 'Smiley', 'SpecialChar', 'PageBreak', 'Iframe' ] },
		'/',
		{ name: 'styles', items: [ 'Styles', 'Format', 'Font', 'FontSize' ] },
		{ name: 'colors', items: [ 'TextColor', 'BGColor' ] },
		{ name: 'tools', items: [ 'Maximize', 'ShowBlocks' ] },
		{ name: 'about', items: [ 'About' ] }
	];
```

**Comment by msamsel** [comment]
@wwalc but it's coherent with other buttons added inside this plugin:
https://github.com/ckeditor/ckeditor-dev/blob/major/plugins/easyimage/plugin.js#L155-L176
```
editor.ui.addButton( 'EasyimageAlt', {
	label: editor.lang.easyimage.commands.altText,
	command: 'easyimageAlt',
	toolbar: 'easyimage,3'
} );
```
```
editor.ui.addButton( 'Easyimage' + capitalize( style ), {
	label: styles[ style ].label,
	command: 'easyimage' + capitalize( style ),
	toolbar: 'easyimage,99',
	icon: styles[ style ].icon,
	iconHiDpi: styles[ style ].iconHiDpi
} );
```

**Comment by msamsel** [comment]
I've remove `file` from names.

File: plugins/easyimage/plugin.js [hunk 1]
```diff
@@ -504,10 +504,50 @@
 		return !CKEDITOR.env.ie || CKEDITOR.env.version >= 11;
 	}
 
+	function addUploadButtonToToolbar( editor ) {
+		var hiddenUploadElement;
+		if ( !editor._.easyImageHiddenUploadElement ) {
+			// Element is not attached to DOM, but still it might be `virtually` clicked.
+			hiddenUploadElement = CKEDITOR.dom.element.createFromHtml( '<input data-cke-easyimage-hidden-upload="1" style="display:none;" type="file" accept="image/*" tabindex="-1" multiple="true">' );
```
Comment (Comandeer) [inline plugins/easyimage/plugin.js]: Most of these attributes are not needed. `[type]`, `[accept]` and `[multiple]` should be enough.

What's more, value for [`[multiple]` attribute](https://html.spec.whatwg.org/#the-multiple-attribute) is incorrect. It's a boolean attribute, so it shouldn't have a value or (if we prefer XHTML style) it's value should be set to the attribute's name (`[multiple=multiple]`).

File: plugins/easyimage/plugin.js [hunk 2]
```diff
@@ -504,10 +504,50 @@
...
+					editor.fire( 'paste', {
+						method: 'paste',
+						dataValue: '',
+						dataTransfer: new CKEDITOR.plugins.clipboard.dataTransfer( { files: evt.target.files } )
+					} );
+				}
+			} );
+
+			editor.on( 'destroy', function() {
+				hiddenUploadElement.removeAllListeners();
```
Comment (Comandeer) [inline plugins/easyimage/plugin.js]: This function does nothing if listeners are bound via native DOM `addEventListener` as our editor does not hold any references to them.

**Comment by mlewand** [comment]
@msamsel @Comandeer please hold up with the rebase, we'll have a delay in shipping 4.9 (next ETA is at the beginning of the next week) we could try to put it still into 4.9.

File: plugins/easyimage/plugin.js [hunk 3]
```diff
@@ -504,10 +504,50 @@
 		return !CKEDITOR.env.ie || CKEDITOR.env.version >= 11;
 	}
 
+	function addUploadButtonToToolbar( editor ) {
+		var hiddenUploadElement;
+		if ( !editor._.easyImageHiddenUploadElement ) {
+			// Element is not attached to DOM, but still it might be `virtually` clicked.
+			hiddenUploadElement = CKEDITOR.dom.element.createFromHtml( '<input data-cke-easyimage-hidden-upload="1" style="display:none;" type="file" accept="image/*" tabindex="-1" multiple="true">' );
+
+			hiddenUploadElement.$.addEventListener( 'change', function( evt ) {
```
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/easyimage/plugin.js]: I'm not sure what you have on your mind with

> "It's also worth trying to swap `change` event for `input` one."

But I hope that using our event handling mechanism will be enough, if not, then I need some clarification how to interpret it.

**Comment by msamsel** [comment]
@Comandeer @mlewand if this feature will be implemented with 4.9.0, then manual test tags should be updated, and changelog entry should be removed from this PR.

**Comment by wwalc** [comment]
> @wwalc but it's coherent with other buttons added inside this plugin:
> https://github.com/ckeditor/ckeditor-dev/blob/major/plugins/easyimage/plugin.js#L155-L176

It just means that the other buttons introduced in this plugin are using wrong convention as well.

**Comment by mlewand** [comment]
Since #1755 got in today it means `easyImageTools.getToken` no longer exists. Also config props `cloudServices_url` and `cloudServices_token` were renamed/repurposed. Check other manual tests for a reference.

**Comment by msamsel** [comment]
@Comandeer 
- branch is rebased
- changelog entry is removed
- manual test has updated tags and methods which initialise cloud services

**Comment by mlewand** [comment]
> Just $0.02: EasyimageUploadFile button name looks a bit invalid. The feature supports uploading images, not files. How about EasyImageUpload? Or just EasyImage?

Extracted to #1769.

**Pull request merged**